Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -158,42 +130,93 @@ public class XPCClient {
+        ///
+        /// It is a programming error to specify this type and provide `forService(named:ofType:)` a name which does not correspond to an XPC service
+        /// bundled with the calling app. You may find this behavior helpful for debugging purposes.
+        case xpcService
     }
     
-    /// Provides a client to communicate with an XPC Mach service.
-    ///
-    /// XPC Mach services are often used by tools such as Launch Agents, Launch Daemons, and helper tools installed with
-    /// [  `SMJobBless`](https://developer.apple.com/documentation/servicemanagement/1431078-smjobbless).
+    /// Provides a client to communicate with a service.
     ///
-    /// In order for this client to be able to communicate with the tool, the tool itself must retrieve and configure an ``XPCServer`` by calling
-    /// ``XPCServer/forThisMachService(named:clientRequirements:)`` or ``XPCServer/forThisBlessedHelperTool()``.
+    /// In order for this client to be able to communicate with the service, the service itself must retrieve and configure an ``XPCServer`` by calling
+    /// ``XPCServer/forThisProcess(ofType:)``.
     ///
-    /// > Note: Client creation always succeeds regardless of whether the XPC Mach service actually exists.
+    /// > Note: Client creation can succeed regardless of whether the service actually exists.
     ///
     /// - Parameters:
-    ///    - named: A key in the `MachServices` entry of the tool's launchd property list.
+    ///   - named: The `CFBundleIdentifier` of the XPC service or the name of the XPC Mach service. For most Mach services the name is specified
+    ///            with the `MachServices` launchd property list entry; however, for login items the name is its`CFBundleIdentifier`.
+    ///   - ofType: There are multiple different types of XPC clients and normally you do not need to concern yourself with this. However, if you are trying to
+    ///             create a client for an XPC Mach service *and* you have an XPC service with a CFBundleIdentifier value that's the same as the name of
+    ///             that Mach service, then you must call this function and explicitly set the type to ``ServiceType/machService``. This is because auto
+    ///             detection will always choose the XPC service if one exists with the provided name.
     /// - Returns: A client configured to communicate with the named service.
-    public static func forMachService(named machServiceName: String) -> XPCClient {
-        XPCMachClient(machServiceName: machServiceName)
+    public static func forService(named serviceName: String, ofType type: ServiceType = .autoDetect) -> XPCClient {
+        switch type {
+            case .autoDetect:
+                if bundledXPCServiceIdentifiers.contains(serviceName) {
+                    return XPCServiceClient(xpcServiceName: serviceName)
+                } else {
+                    return XPCMachClient(machServiceName: serviceName)
+                }
```

Comment:
I'd like to auto-detect this because it means the caller doesn't need to know whether their service/helper is an XPC service or an XPC Mach service. It's not necessarily obvious to the caller what "type" of XPC needs to be used, for example a login item implicitly creates an XPC Mach service using its bundle identifier. By having auto detection for both the client and server, the caller only needs to know the name of the service they're connecting to (and I have very rough ideas of further simplifying that in the future with a Swift package plugin).

How is this a leaky abstraction?

Reply:
> for example a login item implicitly creates an XPC Mach service using its bundle identifier.

Ah I didn't know this, I've never worked with login items

> How is this a leaky abstraction?

Apart from the login item use case, you either have an XPC service or a Mach service. You would know which of the two you have, because you would have had to go through the steps of making one or the other.

Reply:
> Apart from the login item use case, you either have an XPC service or a Mach service. You would know which of the two you have, because you would have had to go through the steps of making one or the other.


While this is true for numerous cases, what's leaking?

To be clear the login item was given as an example, but I don't believe it's the only case.

For an [Endpoint Security](https://developer.apple.com/documentation/endpointsecurity) "a default mach service (name: <teamID>.<bundleID>.xpc) will be provided" (although a Mach service name can also be explicitly set with `NSEndpointSecurityMachServiceName`). [Network Extensions](https://developer.apple.com/documentation/networkextension/filtering_network_traffic) do have the user specify an XPC Service name but it's by add a `NEMachServiceName` key inside of the `NetworkExtension` key. While that does say "Mach", it's still different than the `MachServices` key used by a daemon for example. (FWIW I'd eventually like to add built-in support for both of these to SecureXPC as well.)

The additional idea I had as a potential follow up was to introduce a non-default `autoDetectBundledOnly` which `fatalError`s if the name of the service can't be found anywhere within the bundle, my hope is this would help with debugging when people typo service names or altogether fail to configure them.

Reply:
> For an [Endpoint Security](https://developer.apple.com/documentation/endpointsecurity) "a default mach service (name: <teamID>.<bundleID>.xpc) will be provided" (although a Mach service name can also be explicitly set with NSEndpointSecurityMachServiceName). [Network Extensions](https://developer.apple.com/documentation/networkextension/filtering_network_traffic) do have the user specify an XPC Service name but it's by add a NEMachServiceName key inside of the NetworkExtension key. While that does say "Mach", it's still different than the MachServices key used by a daemon for example. (FWIW I'd eventually like to add built-in support for both of these to SecureXPC as well.)

TIL that these avenues for creating mach services exist. Didn't know!

Reply:
> my hope is this would help with debugging when people typo service names or altogether fail to configure them.

Yes please.

By far the biggest pain point with XPC (in my experience) is what a nightmare it is to debug, and what steaming hot garbage the error messages are. I think we've done a really great job here and making that more sane by "reading the tea leaves" on behalf of users, and I support every effort that advances that! :D